e3sm_to_cmip/argparser.py

```diff
@@ -0,0 +1,300 @@
+        help=(
+            "The path to a json file with additional custom metadata to add to "
+            "the output files."
+        ),
+    )
+    optional_info.add_argument(
+        "--info-out",
+        type=str,
+        help=(
+            "If passed with the --info flag, will cause the variable info to be "
+            "written out to the specified file path as yaml."
+        ),
+    )
+
+    # ======================================================================
+    # Helper arguments.
+    # ======================================================================
+    helper.add_argument(
+        "-h",
+        "--help",
+        action="help",
+        default=argparse.SUPPRESS,
+        help="show this help message and exit",
+    )
+
+    helper.add_argument(
+        "--version",
+        help="Print the version number and exit.",
+        action="version",
+        version=f"%(prog)s {__version__}",
+    )
+
+    return parser
+
+
+def parse_args(args: List[str] | None) -> argparse.Namespace:
+    """Parses command line arguments.
+
+    Parameters
+    ----------
+    args : List[str] | None
+        A list of arguments, useful for debugging purposes to simulate a
+        passing arguments via the CLI.
+
+    Returns
+    -------
+    CLIArguments
+        A data class of parsed arguments.
+    """
+    argparser = setup_argparser()
+
+    try:
+        args_to_parse = sys.argv[1:] if args is None else args
+    except (Exception, BaseException):
+        argparser.print_help()
+        sys.exit(1)
+
+    # Parse the arguments and perform validation.
+    parsed_args = argparser.parse_args(args_to_parse)
+    _validate_parsed_args(parsed_args)
+
+    return parsed_args
+
+
+def _validate_parsed_args(parsed_args: argparse.Namespace):
+    if parsed_args.realm == "mpaso" and not parsed_args.map:
+        raise ValueError("MPAS ocean handling requires a map file")
+
+    if parsed_args.realm == "mpassi" and not parsed_args.map:
+        raise ValueError("MPAS sea-ice handling requires a map file")
+
+    if not parsed_args.simple and not parsed_args.tables_path and not parsed_args.info:
+        raise ValueError("Running without the --simple flag requires CMIP6 tables path")
+
+    if (
+        not parsed_args.input_path or not parsed_args.output_path
+    ) and not parsed_args.info:
+        raise ValueError("Input and output paths required")
+
+    if (
+        not parsed_args.simple
+        and not parsed_args.user_metadata
+        and not parsed_args.info
+    ):
+        raise ValueError(
+            "Running without the --simple flag requires CMIP6 metadata json file"
+        )
+
+    valid_freqs = [freq for freq_type in FREQUENCIES.values() for freq in freq_type]
+    if parsed_args.freq and parsed_args.freq not in valid_freqs:
+        raise ValueError(
+            f"Frequency set to {parsed_args.freq} which is not in the set of allowed "
+            "frequencies: {', '.join(valid_freqs)}"
+        )
```

Comment:
`argparse.py` file of interest

Reply:
@tomvothecoder  This looks excellent - although  I have a hard time seeing how the "flow" for "--help" lead to an exit when encountered.  I trust I will see it in the full-code.